### PR TITLE
Module 8 IP Filtering

### PR DESCRIPTION
**Pull request by Shibbaz**
## Describe your changes
IP Filtering middleware

## Checklist before requesting a review
- [x] I have performed a self-review of my code
- [x] I have added tests and they are passing
- [x] I have followed linters rules(rubocop)

### CODE REVIEW

**Comment by Shibbaz** [comment]
Ready for review.

File: Gemfile [hunk 1]
```diff
@@ -58,28 +58,30 @@ group :development, :test do
   gem 'pry', '~> 0.13.1'
   gem 'rspec'
   gem 'rspec-core'
-  gem 'rspec-rails', '~> 6.0.0.rc1'
```
Comment (DenisNikolski) [inline Gemfile:61]: you should leave rspec gem here, it needs to be in test env as well

**Review by DenisNikolski — COMMENTED** [review]
this PR contains commits from module 6, 
please change base branch to `Module-6`

File: Gemfile (hunk 1 above)
Comment (Shibbaz) [reply Gemfile:61]: @DenisNikolski That's the problem with working on not approved features and checkout -b from branches.

File: spec/requests/movies/delete_spec.rb [hunk 2]
```diff
@@ -55,4 +55,30 @@
       end
     end
   end
+
+  describe 'Stub Remote Ip' do
```
Comment (DenisNikolski) [inline spec/requests/movies/delete_spec.rb:59]: let's be DRY and don't create such specs for each endpoint, 

you can just create one spec, like `spec/requests/ip_filtering_spec.rb` and have in it these contexts, where response is forbidden and where it is not

File: spec/requests/movies/delete_spec.rb [hunk 3]
```diff
@@ -55,4 +55,30 @@
...
+      context 'when valid params' do
+        let(:movie) { create(:movie) }
+
+        let(:params) { { id: movie.id } }
+
+        it 'deletes a record' do
+          delete "/movies/#{movie.id}", params:, env: { "REMOTE_ADDR": '168.121.1.1' }, as: :html
+          expect(response.body).to eq('Your IP is not on IP While List!')
+          expect(response.status).to eq(403)
+        end
```
Comment (DenisNikolski) [inline spec/requests/movies/delete_spec.rb:70]: exceptions is the same as for negative case, does it works as expected?

File: spec/requests/ip_filtering_spec.rb [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+
+    describe 'GET /users' do
+      context 'when there are records' do
+        before { create_list(:user, 5) }
+
+        it 'is forbiden to send a Request' do
+          get '/users', params: {}, env: { "REMOTE_ADDR": '168.121.1.1' }, as: :html
+          expect(response.body).to eq('Your IP is not on IP While List!')
+          expect(response.status).to eq(403)
+        end
```
Comment (DenisNikolski) [inline spec/requests/ip_filtering_spec.rb:25]: what's the difference between this two contexts?  
I would expect one positive and one negative contexts

File: spec/requests/movies/put_spec.rb [hunk 5]
```diff
@@ -43,17 +43,6 @@
...
-      context 'when not valid params' do
-        subject { put '/movies/1000000', params:, as: :html }
-
-        let(:params) { { id: 1_000_000 } }
-
-        it 'does not update a record' do
-          expect(subject).to render_template(:edit)
-          expect(response).to have_http_status(:unprocessable_entity)
-        end
-      end
```
Comment (DenisNikolski) [inline spec/requests/movies/put_spec.rb:56]: why is this removed?

File: Gemfile [hunk 6]
```diff
@@ -74,7 +74,6 @@ group :development do
 
   # Speed up commands on slow machines / big apps [https://github.com/rails/spring]
   # gem "spring"
-  gem 'rspec-rails', '~> 6.0.0.rc1'
```
Comment (DenisNikolski) [inline Gemfile:77]: you can't run tests without rspec-rails

Comment (Shibbaz) [reply Gemfile:77]: @DenisNikolski It's in group :development, :test so It will work.
We don't need rspec on production env.

File: spec/requests/movies/put_spec.rb (hunk 5 above)
Comment (Shibbaz) [reply spec/requests/movies/put_spec.rb:56]: @DenisNikolski moving only get requests to one file spec ip_filtering_spec.rb.

File: Gemfile [hunk 7]
```diff
@@ -85,4 +84,5 @@ group :test do
   gem 'simplecov', require: false
   gem 'webdrivers'
 end
+gem 'dotenv-rails', groups: %i[development test]
```
Comment (DenisNikolski) [inline Gemfile:87]:
```suggestion
gem 'dotenv-rails'
```
it will be useful for all envs

File: spec/requests/movies/put_spec.rb (hunk 5 above)
Comment (DenisNikolski) [reply spec/requests/movies/put_spec.rb:56]: i don't get it, this context is about unprocessable_entity, not forbidden error

Comment (Shibbaz) [reply spec/requests/movies/put_spec.rb:56]: It does not update, if you look into movies def update you can see that the status when failing request.

Comment (DenisNikolski) [reply spec/requests/movies/put_spec.rb:56]: okay, why you delete it?

Comment (DenisNikolski) [reply spec/requests/movies/put_spec.rb:56]: this is `put` request, 
` ip_filtering_spec.rb` contains `get`

Comment (Shibbaz) [reply spec/requests/movies/put_spec.rb:56]: Accidentally when removing ip stub specs.

**Review by DenisNikolski — APPROVED** [review]
